src/librustc/mir/mod.rs

```diff
@@ -1253,6 +1253,15 @@ pub enum StatementKind<'tcx> {
     /// (The starting point(s) arise implicitly from borrows.)
     EndRegion(region::Scope),
 
+    /// Encodes a user's type assertion. These need to be preserved intact so that NLL can respect
+    /// them. For example:
+    ///
+    ///     let (a, b): (T, U) = y;
+    ///
+    /// Here we would insert a `UserAssertTy<(T, U)>(y)` instruction to check that the type of `y`
+    /// is the right thing.
```

Comment:
It is worth adding to the docs that the `Canonical` is used to capture "inference variables" from user's types.

For example:

```
let x: Vec<_> = ...;
let y: &u32 = ...;
```

would result in `Vec<?0>` and `&'?0 u32` respectively (where `?0` is a canonicalized variable).

Reply:
This has been added but it is slightly below the reviewed line so GitHub hasn't noticed.